nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -517,9 +522,16 @@ in
               done
               poolImported "${pool}" || poolImport "${pool}"  # Try one last time, e.g. to import a degraded pool.
               if poolImported "${pool}"; then
-                ${optionalString cfgZfs.requestEncryptionCredentials ''
+                ${optionalString (if isBool cfgZfs.requestEncryptionCredentials
+                                  then cfgZfs.requestEncryptionCredentials
+                                  else cfgZfs.requestEncryptionCredentials != []) ''
                   ${packages.zfsUser}/sbin/zfs list -rHo name,keylocation ${pool} | while IFS=$'\t' read ds kl; do
-                    (case "$kl" in
+                    (${optionalString (!isBool cfgZfs.requestEncryptionCredentials) ''
+                         if ! echo '${concatStringsSep "\n" cfgZfs.requestEncryptionCredentials}' | grep -qFx "$ds"; then
```

Comment:
I just realized another corner case that I have fixed now.  We need the `-F` flag for `grep` to not interpret the string as a regex because we only want exact matches.  It is unlikely that users have `.` or other special characters in their dataset names, but they are [allowed by ZFS](https://docs.oracle.com/cd/E26505_01/html/E37384/gbcpt.html) and we should account for that.

Reply:
I feel like this shouldn't be done with grep like this. Instead, it seems like we ought to have two different code blocks: One for if it's a bool, and one for if it's a list. If it's a bool, then use the old code. If it's a list, then pluck all the elements of the list whose names begin with `"${pool}/"` and put `load-key` commands for each of them. This can be done at Nix eval time without annoying bash-isms, and doesn't require regex'ing every dataset against every list element on boot.

Reply:
agreed the pool name is known here. so the comparison can be done in nix itself.

Reply:
After thinking a bit about it using `grep` here is the method that requires the least amount of code.